src/mango/src/mango_cursor.erl

```diff
@@ -146,46 +146,86 @@ group_indexes_by_type(Indexes) ->
-                        [ddoc_name(DesignId), ViewName]);
-                _ ->
-                    NoIndexWarning
-            end
-    end,
 
-    maybe_add_warning_int(UseIndexInvalidWarning, UserFun, UserAcc).
+% warn if user specified an index which doesn't exist or isn't valid
+% for the selector.
+% In this scenario, Mango will ignore the index hint and auto-select an index.
+invalid_index_warning(Index, Opts) ->
+    UseIndex = lists:keyfind(use_index, 1, Opts),
+    invalid_index_warning_int(Index, UseIndex).
 
 
-maybe_add_warning_int(ok, _, UserAcc) ->
-   UserAcc;
+invalid_index_warning_int(Index, {use_index, [DesignId]}) ->
+    case filter_indexes([Index], DesignId) of
+        [] ->
+            couch_stats:increment_counter([mango, query_invalid_index]),
+            Reason = fmt("_design/~s was not used because it does not contain a valid index for this query.",
+                [ddoc_name(DesignId)]),
+            [Reason];
+        _ ->
+            []
+    end;
+invalid_index_warning_int(Index, {use_index, [DesignId, ViewName]}) ->
+    case filter_indexes([Index], DesignId, ViewName) of
+        [] ->
+            couch_stats:increment_counter([mango, query_invalid_index]),
+            Reason = fmt("_design/~s, ~s was not used because it is not a valid index for this query.",
+                [ddoc_name(DesignId), ViewName]),
+            [Reason];
+        _ ->
+            []
+    end;
+invalid_index_warning_int(_, _) ->
+    [].
+
+
+% warn if a large number of documents needed to be scanned per result
+% returned, implying a lot of in-memory filtering
+index_scan_warning(#execution_stats {
+                    totalDocsExamined = Docs,
+                    totalQuorumDocsExamined = DocsQuorum,
+                    resultsReturned = ResultCount
+                }) ->
+    % Docs and DocsQuorum are mutually exclusive so it's safe to sum them
+    DocsScanned = Docs + DocsQuorum,
+    Ratio = calculate_index_scan_ratio(DocsScanned, ResultCount),
+    Threshold = config:get_integer("mango", "index_scan_warning_threshold", 10),
+    case Threshold > 0 andalso Ratio > Threshold of
+        true ->
+            couch_stats:increment_counter([mango, too_many_docs_scanned]),
+            Reason = <<"The number of documents examined is high in proportion to the number of results returned. Consider adding a more specific index to improve this.">>,
+            [Reason];
+        false -> []
+    end.
 
-maybe_add_warning_int(Warning, UserFun, UserAcc) ->
-    couch_stats:increment_counter([mango, unindexed_queries]),
-    Arg = {add_key, warning, Warning},
-    {_Go, UserAcc0} = UserFun(Arg, UserAcc),
-    UserAcc0.
+
+calculate_index_scan_ratio(DocsScanned, 0) ->
+    DocsScanned;
+calculate_index_scan_ratio(DocsScanned, ResultCount) ->
+    DocsScanned / ResultCount.
+
+
+maybe_add_warning(UserFun, #cursor{index = Index, opts = Opts}, Stats, UserAcc) ->
```

Comment:
Can you move this to the top so it follows emilio https://github.com/cloudant-labs/emilio/blob/master/priv/documentation/420/description.md